src/hammer-vlsi/hammer_vlsi/hammer_tool.py

```diff
@@ -1064,6 +1064,206 @@ def get_bumps(self) -> Optional[BumpsDefinition]:
+        Call this from any custom hook (not used by any default flow).
+        Note: visualizations generated within a PCBDeliverableTool are mirrored about the y-axis (assumption: flip-chip on PCB)
+        """
+        viz_mode = self.get_setting("vlsi.inputs.visualization.mode") #type: str
+        viz_file = self.get_setting("vlsi.inputs.visualization.svg_file") #type: str
+        shorten_path_depth = self.get_setting("vlsi.inputs.visualization.shorten_path_depth") #type: int
+
+        # Checks
+        if not viz_mode in ["all", "floorplan", "bumps", "footprint"]:
+            self.logger.error("Invalid type for 'vlsi.inputs.visualization.type'! No visualization generated.")
+            return
+
+        fp_consts = self.get_placement_constraints() #type: List[PlacementConstraint]
+        # TODO: make pcb action recognize toplevel constraint if par action had hierarchical constraints
+        try:
+            # Get toplevel constraint
+            top = next(filter(lambda x: x.type == PlacementConstraintType.TopLevel, fp_consts)) #type: PlacementConstraint
+        except:
+            self.logger.error("You must at least have a 'toplevel' type placement constraint to generate a visualization!")
+            return
+
+        # TODO: will break if an alternate PCB tool is implemented without naming_scheme property.
+        # TODO: support fractional bump coordinates when generating footprint
+        from .hammer_vlsi_impl import HammerPCBDeliverableTool
+        is_pcb_tool = isinstance(self, HammerPCBDeliverableTool)
+        if not is_pcb_tool:
+            if viz_mode == "footprint":
+                self.logger.error("Can't generate footprint alone if this isn't a PCBDeliverableTool.")
+                return
+            elif viz_mode == "all":
+                self.logger.info("Visualizing in 'all' mode but not a PCBDeliverableTool. Skipping drawing pad designators.")
+
+        # Shorten path names to first letters except lowest "depth" levels in hierarchy
+        def shorten(path: str) -> str:
+            if shorten_path_depth > 0:
+                parent = path.split('/')[:-shorten_path_depth]
+                inst = path.split('/')[-shorten_path_depth:]
+                return os.path.join(*[p[0] for p in parent], *inst)
+            else:
+                return path
+
+        # Get all bump & macro sizes from their masters when constraint width & height are not defined
+        def get_macro_wh(macro: Optional[str]) -> Tuple[Decimal, Decimal]:
+            # take 1st definition of macro
+            try:
+                size = next(filter(lambda x: x.name == macro, macro_sizes))
+                return (size.width, size.height)
+            except:
+                self.logger.warning("Size of {} not found in any LEFs! Defaulting to 10um x 10um. Try manually specifying width & height.".format(macro))
+                return (Decimal("10"), Decimal("10"))
```

Comment:
Is this a common issue that we want to support?
I would have assumed you just want to error.

Reply:
No, I don't think so... not yet at least because we don't check for sizes anywhere else yet (will need to when Chisel floorplanning stuff matures). I don't want to error because it's more obvious that you probably left out an `extra_libraries` entry or something when seeing the viz with the 10x10 block rather than just seeing the text of the error, which gets lost in the flurry of other logger warnings during the `get_macro_sizes()` call.